service/s3/s3manager/download.go

```diff
@@ -154,39 +154,44 @@ func (d *downloader) init() {
 func (d *downloader) download() (n int64, err error) {
 	d.init()
 
-	// Spin up workers
-	ch := make(chan dlchunk, d.ctx.Concurrency)
-	for i := 0; i < d.ctx.Concurrency; i++ {
-		d.wg.Add(1)
-		go d.downloadPart(ch)
-	}
+	// Spin off first worker to check additional header information
+	d.getChunk()
+
+	if total := d.getTotalBytes(); total >= 0 {
+		// Spin up workers
+		ch := make(chan dlchunk, d.ctx.Concurrency)
+
+		for i := 0; i < d.ctx.Concurrency; i++ {
+			d.wg.Add(1)
+			go d.downloadPart(ch)
+		}
 
-	// Assign work
-	for d.geterr() == nil {
-		if d.pos != 0 {
-			// This is not the first chunk, let's wait until we know the total
-			// size of the payload so we can see if we have read the entire
-			// object.
-			total := d.getTotalBytes()
-
-			if total < 0 {
-				// Total has not yet been set, so sleep and loop around while
-				// waiting for our first worker to resolve this value.
-				time.Sleep(10 * time.Millisecond)
-				continue
-			} else if d.pos >= total {
+		// Assign work
+		for d.getErr() == nil {
+			if d.pos >= total {
 				break // We're finished queueing chunks
 			}
+
+			// Queue the next range of bytes to read.
+			ch <- dlchunk{w: d.w, start: d.pos, size: d.ctx.PartSize}
+			d.pos += d.ctx.PartSize
 		}
 
-		// Queue the next range of bytes to read.
-		ch <- dlchunk{w: d.w, start: d.pos, size: d.ctx.PartSize}
-		d.pos += d.ctx.PartSize
-	}
+		// Wait for completion
+		close(ch)
+		d.wg.Wait()
+	} else {
+		// Checking if we read anything new
+		for d.err == nil {
+			d.getChunk()
+		}
 
-	// Wait for completion
-	close(ch)
-	d.wg.Wait()
+		// We expect a 416 error letting us know we are done downloading the
+		// total bytes
+		if d.err.(awserr.RequestFailure).StatusCode() == 416 {
```

Comment:
This should use a runtime type assertion or can panic.

```go
if e, ok := d.err.(awserr.RequestFailure); ok && e.StatusCode() == 416
```

And replace the 416 with `http.StatusRequestedRangeNotSatisfiable` to easier to read